File: frontend/src/mocks/handlers/scan.ts
/**
 * PII scan API handlers
 *
 * Supports all th.scan() parameters:
 * - columns: specific columns to scan
 * - regulations: privacy regulations to check (gdpr, ccpa, lgpd)
 * - min_confidence: confidence threshold for PII detection
 */

import { http, HttpResponse, delay } from 'msw'
import { getStore, getById, create, getAll } from '../data/store'
import { createPIIScan, createId } from '../factories'

const API_BASE = '/api/v1'

/**
 * PII scan request options (mirrors PIIScanRequest from API)
 */
interface PIIScanRequest {
  columns?: string[]
  regulations?: ('gdpr' | 'ccpa' | 'lgpd')[]
  min_confidence?: number
}

export const scanHandlers = [
  // Run PII scan for a source
  http.post(`${API_BASE}/scans/sources/:sourceId/scan`, async ({ params, request }) => {
    await delay(1000) // Simulate scan time

    const sourceId = params.sourceId as string
    const source = getById(getStore().sources, sourceId)

    if (!source) {
      return HttpResponse.json(
        { detail: 'Source not found' },
        { status: 404 }
      )
    }

    // Parse request body for scan options
    let options: PIIScanRequest = {}
    try {
      const body = await request.json()
      options = body as PIIScanRequest
    } catch {
      // Empty body is valid - use defaults
    }

    // Create new PII scan with options
    const scan = createPIIScan({
      id: createId(),
      sourceId,
      options: {
        columns: options.columns,
        regulations: options.regulations,
        min_confidence: options.min_confidence,
      },
    })

    create(getStore().piiScans, scan)

    return HttpResponse.json(scan)
  }),

  // Get PII scan by ID
  http.get(`${API_BASE}/scans/:id`, async ({ params }) => {
    await delay(150)

    const scan = getById(getStore().piiScans, params.id as string)

    if (!scan) {
      return HttpResponse.json(
        { detail: 'PII scan not found' },
        { status: 404 }
      )
    }

    return HttpResponse.json(scan)
  }),

  // List PII scans for a source
  http.get(`${API_BASE}/scans/sources/:sourceId/scans`, async ({ params, request }) => {
    await delay(200)

    const url = new URL(request.url)
    const limit = parseInt(url.searchParams.get('limit') ?? '20')

    const sourceId = params.sourceId as string
    const source = getById(getStore().sources, sourceId)

    if (!source) {
      return HttpResponse.json(
        { detail: 'Source not found' },
        { status: 404 }
      )
    }

    const allScans = getPIIScansBySourceId(sourceId)
    const scans = allScans.slice(0, limit)

    return HttpResponse.json({
      data: scans,
      total: allScans.length,
      limit,
    })
  }),

  // Get latest PII scan for a source
  http.get(`${API_BASE}/scans/sources/:sourceId/scans/latest`, async ({ params }) => {
    await delay(150)

    const sourceId = params.sourceId as string
    const source = getById(getStore().sources, sourceId)

    if (!source) {
      return HttpResponse.json(
        { detail: 'Source not found' },
        { status: 404 }
      )
    }

    const scans = getPIIScansBySourceId(sourceId)

    if (scans.length === 0) {
      return HttpResponse.json(
        { detail: 'No PII scans found for source' },
        { status: 404 }
      )
    }

    return HttpResponse.json(scans[0])
  }),
]

// Helper function to get PII scans by source ID
function getPIIScansBySourceId(sourceId: string) {
  return getAll(getStore().piiScans)
    .filter((s) => s.source_id === sourceId)
    .sort((a, b) => new Date(b.created_at).getTime() - new Date(a.created_at).getTime())
}

File: src/truthound_dashboard/api/scan.py
"""PII scan API endpoints.

This module provides endpoints for running PII scans using th.scan().
Detects personally identifiable information and checks compliance with
privacy regulations (GDPR, CCPA, LGPD).
"""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, HTTPException, Path, Query

from truthound_dashboard.schemas import (
    PIIScanListItem,
    PIIScanListResponse,
    PIIScanRequest,
    PIIScanResponse,
)

from .deps import PIIScanServiceDep, SourceServiceDep

router = APIRouter()


@router.post(
    "/sources/{source_id}/scan",
    response_model=PIIScanResponse,
    summary="Run PII scan",
    description="Scan data source for personally identifiable information (PII)",
)
async def run_pii_scan(
    service: PIIScanServiceDep,
    source_id: Annotated[str, Path(description="Source ID to scan")],
    request: PIIScanRequest,
) -> PIIScanResponse:
    """Run PII scan on a data source.

    Supports all th.scan() parameters for maximum flexibility:
    - columns: Specific columns to scan
    - regulations: Privacy regulations to check (gdpr, ccpa, lgpd)
    - min_confidence: Confidence threshold for PII detection

    Args:
        service: Injected PII scan service.
        source_id: Source to scan.
        request: Scan options.

    Returns:
        PII scan result with findings and violations.

    Raises:
        HTTPException: 404 if source not found.
    """
    try:
        scan = await service.run_scan(
            source_id,
            columns=request.columns,
            regulations=request.regulations,
            min_confidence=request.min_confidence,
        )
        return PIIScanResponse.from_model(scan)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.get(
    "/{scan_id}",
    response_model=PIIScanResponse,
    summary="Get PII scan",
    description="Get a specific PII scan result by ID",
)
async def get_pii_scan(
    service: PIIScanServiceDep,
    scan_id: Annotated[str, Path(description="Scan ID")],
) -> PIIScanResponse:
    """Get a specific PII scan result.

    Args:
        service: Injected PII scan service.
        scan_id: Scan unique identifier.

    Returns:
        PII scan details with findings and violations.

    Raises:
        HTTPException: 404 if scan not found.
    """
    scan = await service.get_scan(scan_id)
    if scan is None:
        raise HTTPException(status_code=404, detail="PII scan not found")
    return PIIScanResponse.from_model(scan)


@router.get(
    "/sources/{source_id}/scans",
    response_model=PIIScanListResponse,
    summary="List source PII scans",
    description="Get PII scan history for a source",
)
async def list_source_pii_scans(
    service: PIIScanServiceDep,
    source_service: SourceServiceDep,
    source_id: Annotated[str, Path(description="Source ID")],
    limit: Annotated[int, Query(ge=1, le=100, description="Maximum items")] = 20,
) -> PIIScanListResponse:
    """List PII scan history for a source.

    Args:
        service: Injected PII scan service.
        source_service: Injected source service.
        source_id: Source to get scans for.
        limit: Maximum scans to return.

    Returns:
        List of PII scan summaries.

    Raises:
        HTTPException: 404 if source not found.
    """
    # Verify source exists
    source = await source_service.get_by_id(source_id)
    if source is None:
        raise HTTPException(status_code=404, detail="Source not found")

    scans = await service.list_for_source(source_id, limit=limit)

    return PIIScanListResponse(
        data=[PIIScanListItem.from_model(s) for s in scans],
        total=len(scans),
        limit=limit,
    )


@router.get(
    "/sources/{source_id}/scans/latest",
    response_model=PIIScanResponse | None,
    summary="Get latest PII scan",
    description="Get the most recent PII scan for a source, or null if no scans exist",
)
async def get_latest_pii_scan(
    service: PIIScanServiceDep,
    source_service: SourceServiceDep,
    source_id: Annotated[str, Path(description="Source ID")],
) -> PIIScanResponse | None:
    """Get the most recent PII scan for a source.

    Args:
        service: Injected PII scan service.
        source_service: Injected source service.
        source_id: Source to get latest scan for.

    Returns:
        Latest PII scan result, or null if no scans exist.

    Raises:
        HTTPException: 404 if source not found.
    """
    # Verify source exists
    source = await source_service.get_by_id(source_id)
    if source is None:
        raise HTTPException(status_code=404, detail="Source not found")

    scan = await service.get_latest_for_source(source_id)
    if scan is None:
        return None

    return PIIScanResponse.from_model(scan)
